src/os/unix/wayland.rs

```diff
@@ -0,0 +1,1117 @@
+
+        Ok((
+            Self {
+                _display: display,
+                wl_display,
+                _compositor: compositor,
+                _base: xdg_wm_base,
+                surface,
+                xdg_surface,
+                toplevel: xdg_toplevel,
+                _shm: shm,
+                event_queue,
+                _seat: seat,
+                xdg_config,
+                cursor,
+                cursor_surface,
+                buf_pool,
+            },
+            input,
+        ))
+    }
+
+    fn set_geometry(&self, pos: (i32, i32), size: (i32, i32)) {
+        self.xdg_surface
+            .set_window_geometry(pos.0, pos.1, size.0, size.1);
+    }
+
+    fn set_title(&self, title: &str) {
+        self.toplevel.set_title(title.to_owned());
+    }
+
+    fn set_no_resize(&self, size: (i32, i32)) {
+        self.toplevel.set_max_size(size.0, size.1);
+        self.toplevel.set_min_size(size.0, size.1);
+    }
+
+    // Sets a specific cursor style
+    fn update_cursor(&mut self, cursor: &str) {
+        let csr = self.cursor.get_cursor(cursor).unwrap();
+        let img = csr.frame_buffer(0).unwrap();
+        self.cursor_surface.attach(Some(&*img), 0, 0);
+        self.cursor_surface.damage(0, 0, 32, 32);
+        self.cursor_surface.commit();
+    }
+
+    // Resizes when buffer is bigger or less
+    fn update_framebuffer(&mut self, buffer: &[u32], size: (i32, i32)) {
+        use std::io::{Seek, SeekFrom};
+
+        let (mut fd, buf) = self.buf_pool.get_buffer(size);
+
+        fd.seek(SeekFrom::Start(0)).unwrap();
+        let slice = unsafe {
+            std::slice::from_raw_parts(
+                buffer[..].as_ptr() as *const u8,
+                buffer.len() * std::mem::size_of::<u32>(),
+            )
+        };
+        fd.write_all(&slice[..]).unwrap();
+        fd.flush().unwrap();
+
+        // Acknowledge the last configure event
+        if let Some(serial) = (*self.xdg_config.borrow_mut()).take() {
+            self.xdg_surface.ack_configure(serial);
+        }
+
+        self.surface.attach(Some(buf), 0, 0);
+        self.surface
+            .damage(0, 0, i32::max_value(), i32::max_value());
+        self.surface.commit();
+    }
+
+    fn get_toplevel_info(&self) -> (ToplevelResolution, ToplevelClosed) {
+        let configure = Rc::new(RefCell::new(None));
+        let close = Rc::new(RefCell::new(false));
+
+        {
+            let configure = configure.clone();
+            let close = close.clone();
+
+            self.toplevel.quick_assign(move |_, event, _| {
+                use wayland_protocols::xdg_shell::client::xdg_toplevel::Event;
+
+                if let Event::Configure { width, height, .. } = event {
+                    *configure.borrow_mut() = Some((width, height));
+                } else if let Event::Close = event {
+                    *close.borrow_mut() = true;
+                }
+            });
+        }
+
+        (configure, close)
```

Comment:
```suggestion
        (resolution, close)
```

Since you named the type `ToplevelResolution`, I think it would make sense to inherit the name here, as well.